src/librustdoc/scrape_examples.rs

```diff
@@ -0,0 +1,208 @@
+
+/// Visitor for traversing a crate and finding instances of function calls.
+struct FindCalls<'a, 'tcx> {
+    tcx: TyCtxt<'tcx>,
+    map: Map<'tcx>,
+    cx: Context<'tcx>,
+    calls: &'a mut AllCallLocations,
+}
+
+crate fn def_id_call_key(tcx: TyCtxt<'_>, def_id: DefId) -> DefIdCallKey {
+    format!(
+        "{}{}",
+        tcx.crate_name(def_id.krate).to_ident_string(),
+        tcx.def_path(def_id).to_string_no_crate_verbose()
+    )
+}
+
+impl<'a, 'tcx> Visitor<'tcx> for FindCalls<'a, 'tcx>
+where
+    'tcx: 'a,
+{
+    type Map = Map<'tcx>;
+
+    fn nested_visit_map(&mut self) -> intravisit::NestedVisitorMap<Self::Map> {
+        intravisit::NestedVisitorMap::OnlyBodies(self.map)
+    }
+
+    fn visit_expr(&mut self, ex: &'tcx hir::Expr<'tcx>) {
+        intravisit::walk_expr(self, ex);
+
+        // Get type of function if expression is a function call
+        let (ty, span) = match ex.kind {
+            hir::ExprKind::Call(f, _) => {
+                let types = self.tcx.typeck(ex.hir_id.owner);
+                (types.node_type(f.hir_id), ex.span)
+            }
+            hir::ExprKind::MethodCall(_, _, _, span) => {
+                let types = self.tcx.typeck(ex.hir_id.owner);
+                let def_id = types.type_dependent_def_id(ex.hir_id).unwrap();
+                (self.tcx.type_of(def_id), span)
+            }
+            _ => {
+                return;
+            }
+        };
+
+        if span.from_expansion() {
+            return;
+        }
+
+        // Save call site if the function resolves to a concrete definition
+        if let ty::FnDef(def_id, _) = ty.kind() {
+            let fn_key = def_id_call_key(self.tcx, *def_id);
+            let entries = self.calls.entry(fn_key).or_insert_with(FxHashMap::default);
+            let file = self.tcx.sess.source_map().lookup_char_pos(span.lo()).file;
+            let file_path = match file.name.clone() {
+                FileName::Real(real_filename) => real_filename.into_local_path(),
+                _ => None,
+            };
+
+            if let Some(file_path) = file_path {
+                let abs_path = fs::canonicalize(file_path.clone()).unwrap();
+                let cx = &self.cx;
+                let location = CallLocation::new(self.tcx, span, ex.hir_id, &file);
+
+                entries
+                    .entry(abs_path)
+                    .or_insert_with(|| {
+                        let clean_span = crate::clean::types::Span::new(span);
+                        let url = cx.href_from_span(clean_span).unwrap();
+                        let display_name = file_path.display().to_string();
+                        CallData { locations: Vec::new(), url, display_name }
+                    })
+                    .locations
+                    .push(location);
+            }
+        }
+    }
+}
+
+crate fn run(
+    krate: clean::Crate,
+    renderopts: config::RenderOptions,
+    cache: formats::cache::Cache,
```

Comment:
So the core issue is that examples are scraped from reverse-dependencies. AFAIK, such files aren't guaranteed to be in the source map while documenting the crate that defines a given function. This has a few implications:
* Spans don't translate from the scraped crate to the documented crate. Hence, the need for a separate `SyntaxRange` structure, and why I have to generate the example URL at scrape time instead of document time.
* DefIds don't translate from crate to crate. The DefPathHash of a function while scraping doesn't seem to be the same as while documenting. This one I'm less sure why, but it's been my observation when trying to use `DefPathHash` to identify a function. Hence I have the hacky `def_id_call_key` function instead.

I don't know cargo or rustc well enough to say whether these issues are fixable.

Reply:
> Spans don't translate from the scraped crate to the documented crate.

Hmm, is that because you don't have the DefIds available? Are these spans of the calls *in the example*? If so that makes more sense to me and I don't think it's fixable.

(If you did have the ids available, I think you could use https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/context/struct.TyCtxt.html#method.def_span.)

> The DefPathHash of a function while scraping doesn't seem to be the same as while documenting.

Can you give an example of a change between the two? It seems very strange to me that the DefPathHash would change, my understanding is it's supposed to be stable across compilations.

Reply:
For the record: yes, these are the spans in the example, which is why they have to be reserialized. Will found that the DefPathHash was different because cargo was passing a different -C metadata parameter; he's going to fix that in the cargo PR.